### PR TITLE
[Flutter GPU] Remove Command/VertexBuffer usage from Flutter GPU.

### PR DESCRIPTION
**Pull request by bdero**
Resolves https://github.com/flutter/flutter/issues/156764.
Resolves https://github.com/flutter/flutter/issues/155335.

- Remove Command/VertexBuffer from Flutter GPU.
- Separate vertex/index count into `impeller::RenderPass::SetElementCount(size_t count)` & accurately document its behavior.
- Make Flutter GPU uniform/texture bindings re-assignable.
- Remove unnecessary templates.

### CODE REVIEW

File: impeller/renderer/render_pass.h [hunk 1]
```diff
@@ -89,6 +89,12 @@ class RenderPass : public ResourceBinder {
   ///
   virtual void SetScissor(IRect scissor);
 
+  //----------------------------------------------------------------------------
+  /// The number of elements to draw. When only a vertex buffer is set, this is
+  /// the vertex count. When an index buffer is set, this is the index count.
+  ///
+  virtual void SetElementCount(size_t count);
```
Comment (bdero) [inline impeller/renderer/render_pass.h:96]: New method in RenderPass here.

File: impeller/renderer/render_pass.h [hunk 2]
```diff
@@ -115,11 +121,9 @@ class RenderPass : public ResourceBinder {
   ///
   /// @param[in]  vertex_buffer  The buffer view to use for sourcing vertices.
   ///
-  /// @param[in]  vertex_count   The number of vertices to draw.
-  ///
   /// @return     Returns false if the given buffer view is invalid.
   ///
-  bool SetVertexBuffer(BufferView vertex_buffer, size_t vertex_count);
+  bool SetVertexBuffer(BufferView vertex_buffer);
```
Comment (bdero) [inline impeller/renderer/render_pass.h:126]: This patch removes `vertex_count` from `SetVertexBuffer`. It gets set via `SetElementCount` now. (This value becomes the index count when an index buffer is present)

File: lib/gpu/render_pass.h [hunk 3]
```diff
@@ -50,42 +52,59 @@ class RenderPass : public RefCountedDartWrappable<RenderPass> {
...
+      std::unordered_map<const flutter::gpu::Shader::UniformBinding*,
+                         impeller::BufferAndUniformSlot>;
+  using TextureUniformMap =
+      std::unordered_map<const flutter::gpu::Shader::TextureBinding*,
+                         impeller::TextureAndSampler>;
+
+  BufferUniformMap vertex_uniform_bindings;
+  TextureUniformMap vertex_texture_bindings;
+  BufferUniformMap fragment_uniform_bindings;
+  TextureUniformMap fragment_texture_bindings;
```
Comment (bdero) [inline lib/gpu/render_pass.h:79]: Uniform bindings are wiped out after each `RenderPass::Draw` call. `Command` was hanging on to this state between draws. This is similar to what `Command` was doing previously, except it also resolves https://github.com/flutter/flutter/issues/155335 by tracking maps with unique binding identities.

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc [hunk 4]
```diff
@@ -415,7 +415,7 @@ std::optional<size_t> BufferBindingsGLES::BindTextures(
     /// If there is a sampler for the texture at the same index, configure the
     /// bound texture using that sampler.
     ///
-    const auto& sampler_gles = SamplerGLES::Cast(*data.sampler);
+    const auto& sampler_gles = SamplerGLES::Cast(**data.sampler);
```
Comment (jonahwilliams) [inline impeller/renderer/backend/gles/buffer_bindings_gles.cc:418]: Does this count as making you a two star programmer? :)

File: lib/gpu/render_pass.cc [hunk 5]
```diff
@@ -360,20 +345,25 @@ void InternalFlutterGpu_RenderPass_BindVertexBufferHost(
...
-                            int offset_in_bytes,
-                            int length_in_bytes,
-                            int index_type,
-                            int index_count) {
-  auto& vertex_buffer = wrapper->GetVertexBuffer();
-  vertex_buffer.index_buffer = impeller::BufferView{
+static void BindIndexBuffer(
+    flutter::gpu::RenderPass* wrapper,
+    const std::shared_ptr<const impeller::DeviceBuffer>& buffer,
+    int offset_in_bytes,
```
Comment (jonahwilliams) [inline lib/gpu/render_pass.cc:351]: It may not matter if these are internal APIs but using signed ints for values that can't be negative is scary :)

**Review by jonahwilliams — APPROVED** [review]
Still LGTM

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc (hunk 4 above)
Comment (bdero) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:418]: One day I'll sneak three stars in and become a legend.

File: lib/gpu/render_pass.cc (hunk 5 above)
Comment (bdero) [reply lib/gpu/render_pass.cc:351]: Is there an alternative approach we should consider exploring with Dart FFI?

Comment (jonahwilliams) [reply lib/gpu/render_pass.cc:351]: dart::ffi methods support signed/unsigned types, and then you validate on the dart side of things. Or else check on the C++ and then cast.

Comment (bdero) [reply lib/gpu/render_pass.cc:351]: Ah yeah we should use those unsigned types, especially for enums. Added an issue here: https://github.com/flutter/flutter/issues/157129

**Pull request merged**